Классификация: превышает ли значение SI медианное значение выборки

In [2]:

import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score

# загрузка данных
X_train = pd.read_csv('/content/X_train.csv')
X_test = pd.read_csv('/content/X_test.csv')

# функция для обучения и оценки
def evaluate_classifier(y_train_path, y_test_path, model, model_name, task_name):
    y_train = pd.read_csv(y_train_path).values.ravel()
    y_test = pd.read_csv(y_test_path).values.ravel()

    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    y_proba = model.predict_proba(X_test)[:, 1]

    print(f"\n{task_name} - {model_name}")
    print(f"  Accuracy: {accuracy_score(y_test, y_pred):.3f}")
    print(f"  F1-score: {f1_score(y_test, y_pred):.3f}")
    print(f"  ROC-AUC:  {roc_auc_score(y_test, y_proba):.3f}")

print("="*60)
print("КЛАССИФИКАЦИЯ SI (выше/ниже медианы)")
print("="*60)

# Logistic Regression
evaluate_classifier(
    'y_train_si_median.csv', 'y_test_si_median.csv',
    LogisticRegression(max_iter=1000, random_state=42),
    "LogisticRegression", "SI"
)

# Random Forest
evaluate_classifier(
    'y_train_si_median.csv', 'y_test_si_median.csv',
    RandomForestClassifier(n_estimators=100, random_state=42),
    "RandomForest", "SI"
)

# XGBoost
evaluate_classifier(
    'y_train_si_median.csv', 'y_test_si_median.csv',
    XGBClassifier(n_estimators=100, random_state=42, eval_metric='logloss'),
    "XGBoost", "SI"
)

КЛАССИФИКАЦИЯ SI (выше/ниже медианы)

SI - LogisticRegression
  Accuracy: 0.529
  F1-score: 0.000
  ROC-AUC:  0.573

SI - RandomForest
  Accuracy: 0.645
  F1-score: 0.639
  ROC-AUC:  0.695

SI - XGBoost
  Accuracy: 0.612
  F1-score: 0.595
  ROC-AUC:  0.665


Вывод:классификация SI (выше/ниже медианы) оказалась более сложной задачей, чем предсказание IC50 и CC50. Ансамблевые методы показали умеренное качество: Random Forest (AUC = 0.695, Accuracy = 64.5%) и XGBoost (AUC = 0.665). Логистическая регрессия полностью провалилась (F1 = 0.000). Это объясняется тем, что селективность (SI = CC50/IC50) является производным показателем, который накапливает ошибки предсказания обеих величин. Несмотря на более низкие метрики, модель может быть полезна для грубой оценки селективности соединений на ранних этапах скрининга.